  # Capstone Project Notebook #

This notebook will primarily be used for the capstone project as part of the Coursera IBM Data Science Course

In [1]:
import pandas as pd
import numpy as np
import requests

## Getting Wiki Data & Cleaning
1. Removing rows that do not have an assigned borough
2. If a cell has a borough but is not assigned a neighborhood, then we set the neighborhood to be the same as the borough (note that this condition is never met in the dataset, so we omit the operation).

In [2]:
df = pd.read_html("https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969", attrs={"class": "wikitable sortable"})[0]
df = df[df['Borough']!='Not assigned'].reset_index(drop=True)
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [3]:
df.shape

(103, 3)

## Merging with Geospatial Data
Given an external datatset containing latitude and longitude geospatial coordinates for each Postal Code, which we denote below as geo_df, we merge the two dataframes

In [4]:
geo_df = pd.read_csv("~/Desktop/Geospatial_Coordinates.csv")
merged_df = pd.merge(df,geo_df, on='Postal Code')
merged_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [5]:
print(merged_df.shape)

(103, 5)


We can be confident that the merging went smoothly (without losing data), because the dataframe shapes are consistent with expectations

## Visualising Latitude & Longitude Data

In [6]:
!conda install -c conda-forge folium=0.5.0 --yes -q
import folium

Solving environment: ...working... done

# All requested packages already installed.



In [7]:
latitude, longitude = [43.741667,-79.373333] ## found manually on wiki page

# create map of Toronto, Canada using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(merged_df['Latitude'], merged_df['Longitude'], merged_df['Borough'], merged_df['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
         [lat, lng],
         radius=5,
         popup=label,
         color='blue',
         fill=True,
         fill_color='#3186cc',
         fill_opacity=0.7,
         parse_html=False).add_to(map_toronto)  
    
map_toronto

## Segmenting Toronto Boroughs

#### Define FourSquare Credentials & Version

In [8]:
CLIENT_ID = 'NCOK3XLOBXPSZAGSGJ5ND3F0MOYUQGHQIMKHXHWFVSOUZWYV' # your Foursquare ID
CLIENT_SECRET = 'UPPMLT05HXXFDSVBYTEKTPRIBRRHFBMHB35REQM1QH2XPI5B' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value
radius=500

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [10]:
%%capture
toronto_venues = getNearbyVenues(names=merged_df['Neighbourhood'],
                                   latitudes=merged_df['Latitude'],
                                   longitudes=merged_df['Longitude']
                                  );

In [11]:
toronto_venues.groupby('Neighbourhood').size()

Neighbourhood
Agincourt                                           4
Alderwood, Long Branch                              7
Bathurst Manor, Wilson Heights, Downsview North    22
Bayview Village                                     4
Bedford Park, Lawrence Manor East                  24
                                                   ..
Willowdale, Willowdale East                        34
Willowdale, Willowdale West                         4
Woburn                                              3
Woodbine Heights                                    9
York Mills West                                     3
Length: 96, dtype: int64

### GroupBy Issue

When we group by neighbourhoods, the length of the output will depend on the locations of the venues that FourSquare returns. Those neighbourhoods that don't return any venues on a particular day will not be present in the dataset.

merged_df has length 103, while toronto_venues.groupby('Neighbourhood') has length 95-96. When we insert KMeans labels, we need to remove labels that are given a label of NaN (these neighbourhoods contribute nothing to the clustering).

In [14]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 269 unique categories.


In [15]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]


In [1]:
toronto_venues

NameError: name 'toronto_venues' is not defined

In [18]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
91                      Willowdale, Willowdale East                0.0   
92                      Willowdale, Willowdale West                0.0   
93                                           Woburn                0.0   
94                                 Woodbine Heights                0.0   
95                                  York Mills West                0.0   

    Airport  Airport Food Court  Airport Lounge  Airport Service  \
0       0.0                 0.0             0.0              0.0   
1       0.0                 0.0             0.0              0.0   
2       0.0                 0.0             0.0              0.0   
3       0.0                 0.0             0.0              0.0   
4       0.0                 0.0             0.0              0.0   
..      ...                 ...             ...              ...   
91      0.0                 0.0             0.0              0.0   
92      0.0                 0.0             0.0              0.0   
93      0.0                 0.0             0.0              0.0   
94      0.0                 0.0             0.0              0.0   
95      0.0                 0.0             0.0              0.0   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0                0.0             0.000000           0.0       0.0  ...   
1                0.0             0.000000           0.0       0.0  ...   
2                0.0             0.000000           0.0       0.0  ...   
3                0.0             0.000000           0.0       0.0  ...   
4                0.0             0.041667           0.0       0.0  ...   
..               ...                  ...           ...       ...  ...   
91               0.0             0.000000           0.0       0.0  ...   
92               0.0             0.000000           0.0       0.0  ...   
93               0.0             0.000000           0.0       0.0  ...   
94               0.0             0.000000           0.0       0.0  ...   
95               0.0             0.000000           0.0       0.0  ...   

    Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                             0.0               0.0     0.000000   
1                             0.0               0.0     0.000000   
2                             0.0               0.0     0.000000   
3                             0.0               0.0     0.000000   
4                             0.0               0.0     0.000000   
..                            ...               ...          ...   
91                            0.0               0.0     0.000000   
92                            0.0               0.0     0.000000   
93                            0.0               0.0     0.000000   
94                            0.0               0.0     0.111111   
95                            0.0               0.0     0.000000   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                0.000000              0.0       0.0        0.0          0.0   
1                0.000000              0.0       0.0        0.0          0.0   
2                0.000000              0.0       0.0        0.0          0.0   
3                0.000000              0.0       0.0        0.0          0.0   
4                0.000000              0.0       0.0        0.0          0.0   
..                    ...              ...       ...        ...          ...   
91               0.029412              0.0       0.0        0.0   

In [19]:
toronto_grouped.shape

(96, 270)

In [20]:
%%capture

num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

In [21]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [61]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt          Skating Rink   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                Lounge        Breakfast Spot  Latin American Restaurant   
1        Sandwich Place                   Gym                Coffee Shop   
2           Coffee Shop   Fried Chicken Joint                Supermarket   
3                  Café                  Bank         Chinese Restaurant   
4           Coffee Shop        Sandwich Place           Greek Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store      Doner Restaurant            Donut Shop   
1          Skating Rink                   Pub      Doner Restaurant   
2      Sushi Restaurant     Mobile Phone Shop         Deli / Bodega   
3   Distribution Center               Dog Run      Doner Restaurant   
4               Butcher      Sushi Restaurant             Juice Bar   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0             Drugstore   Dumpling Restaurant  Eastern European Restaurant  
1                 Diner        Discount Store          Distribution Center  
2           Bridal Shop    Frozen Yogurt Shop    Middle Eastern Restaurant  
3            Donut Shop             Drugstore                  Yoga Studio  
4  Fast Food Restaurant          Liquor Store                         Café

## K-Means Clustering

In [62]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 10
toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 9, 2, 2, 2, 6,
       2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 5, 4, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 6, 0, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 0, 2, 2, 2, 0,
       2, 8, 2, 2, 7, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2,
       0, 2, 0, 2, 2, 1, 2, 0], dtype=int32)

### Note about K-Means
Interesting to note that as of 26/03/21, the vast majority of neighbourhoods exist in one cluster, regardless of the number of clusters employed. This probably means that the clustering criteria, i.e. the 10 most common venues, does not adequately distinguish between the majority of neighbourhoods. Nonetheless, we continue as though it did. (Alternatively, there is a chance that the K-Means is distinguishing appropriately and that the local government has done very well to ensure consistent coverage of venue types throughout the district).

In [63]:
# add clustering labels
try:
    neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
except ValueError:
    pass

toronto_merged = merged_df

## toronto_merged needs to have elements removed if the clustering label is NaN.
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
toronto_merged.dropna(subset = ['Cluster Labels'], inplace=True)
toronto_merged = toronto_merged.astype({"Cluster Labels": int})

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               0                  Park   
1  43.725882 -79.315572               2           Pizza Place   
2  43.654260 -79.360636               2           Coffee Shop   
3  43.718518 -79.464763               2        Clothing Store   
4  43.662301 -79.389494               2           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0       Food & Drink Shop           Yoga Studio  Eastern European Restaurant   
1            Intersection          Hockey Arena        Portuguese Restaurant   
2                    Park                Bakery                          Pub   
3  Furniture / Home Store     Accessories Store        Vietnamese Restaurant   
4        Sushi Restaurant           Yoga Studio                      Theater   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Dog Run      Doner Restaurant            Donut Shop   
1     French Restaurant           Coffee Shop            Donut Shop   
2        Breakfast Spot                  Café               Theater   
3           Coffee Shop              Boutique       Coworking Space   
4         Smoothie Shop         Burrito Place        Sandwich Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Drugstore   Dumpling Restaurant            Escape Room  
1        Discount Store   Distribution Center                Dog Run  
2           Yoga Studio            Shoe Store             Restaurant  
3               Dog Run        Farmers Market     Falafel Restaurant  
4              Beer Bar    Mexican Restaurant                   Café

In [85]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters

In [92]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
0                                           Parkwoods               0   
21                                Caledonia-Fairbanks               0   
35      East Toronto, Broadview North (Old East York)               0   
52                            Willowdale, Newtonbrook               0   
64                                             Weston               0   
66                                    York Mills West               0   
85  Milliken, Agincourt North, Steeles East, L'Amo...               0   
91                                           Rosedale               0   

   1st Most Common Venue 2nd Most Common Venue        3rd Most Common Venue  \
0                   Park     Food & Drink Shop                  Yoga Studio   
21                  Park         Women's Store                         Pool   
35          Intersection                  Park            Convenience Store   
52                  Park           Yoga Studio  Eastern European Restaurant   
64                  Park         Jewelry Store            Convenience Store   
66                  Park     Convenience Store                  Yoga Studio   
85                  Park            Playground                 Intersection   
91                  Park            Playground                        Trail   

          4th Most Common Venue        5th Most Common Venue  \
0   Eastern European Restaurant                      Dog Run   
21                  Yoga Studio          Dumpling Restaurant   
35          Rental Car Location          Dumpling Restaurant   
52          Distribution Center                      Dog Run   
64                  Yoga Studio  Eastern European Restaurant   
66  Eastern European Restaurant                      Dog Run   
85                 Concert Hall                        Diner   
91                  Yoga Studio                    Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0       Doner Restaurant            Donut Shop             Drugstore   
21   Distribution Center               Dog Run      Doner Restaurant   
35   Distribution Center               Dog Run      Doner Restaurant   
52      Doner Restaurant            Donut Shop             Drugstore   
64               Dog Run      Doner Restaurant            Donut Shop   
66      Doner Restaurant            Donut Shop             Drugstore   
85    Falafel Restaurant           Event Space  Ethiopian Restaurant   
91        Discount Store   Distribution Center               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
0    Dumpling Restaurant            Escape Room  
21            Donut Shop              Drugstore  
35            Donut Shop              Drugstore  
52   Dumpling Restaurant      Electronics Store  
64             Drugstore    Dumpling Restaurant  
66   Dumpling Restaurant            Escape Room  
85           Escape Room      Electronics Store  
91      Doner Restaurant             Donut Shop

In [93]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
22        Woburn               1           Coffee Shop  Korean BBQ Restaurant   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
22           Yoga Studio  Eastern European Restaurant               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22      Doner Restaurant            Donut Shop             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
22   Dumpling Restaurant      Electronics Store

In [94]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
1                                     Victoria Village               2   
2                            Regent Park, Harbourfront               2   
3                     Lawrence Manor, Lawrence Heights               2   
4          Queen's Park, Ontario Provincial Government               2   
7                                            Don Mills               2   
..                                                 ...             ...   
97              First Canadian Place, Underground city               2   
98       The Kingsway, Montgomery Road, Old Mill North               2   
99                                Church and Wellesley               2   
100  Business reply mail Processing Centre, South C...               2   
102  Mimico NW, The Queensway West, South of Bloor,...               2   

    1st Most Common Venue   2nd Most Common Venue 3rd Most Common Venue  \
1             Pizza Place            Intersection          Hockey Arena   
2             Coffee Shop                    Park                Bakery   
3          Clothing Store  Furniture / Home Store     Accessories Store   
4             Coffee Shop        Sushi Restaurant           Yoga Studio   
7                     Gym              Restaurant           Coffee Shop   
..                    ...                     ...                   ...   
97            Coffee Shop                    Café                 Hotel   
98                   Pool                   River           Yoga Studio   
99            Coffee Shop        Sushi Restaurant   Japanese Restaurant   
100           Yoga Studio           Auto Workshop         Garden Center   
102        Hardware Store           Tanning Salon          Burger Joint   

     4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
1    Portuguese Restaurant     French Restaurant           Coffee Shop   
2                      Pub        Breakfast Spot                  Café   
3    Vietnamese Restaurant           Coffee Shop              Boutique   
4                  Theater         Smoothie Shop         Burrito Place   
7     Caribbean Restaurant        Clothing Store   Japanese Restaurant   
..                     ...                   ...                   ...   
97                     Gym            Restaurant   Japanese Restaurant   
98               Drugstore        Discount Store   Distribution Center   
99                 Gay Bar            Restaurant           Yoga Studio   
100                 Garden    Light Rail Station  Fast Food Restaurant   
102             Kids Store        Sandwich Place       Supplement Shop   

    7th Most Common Venue 8th Most Common Venue     9th Most Common Venue  \
1              Donut Shop        Discount Store       Distribution Center   
2                 Theater           Yoga Studio                Shoe Store   
3         Coworking Space               Dog Run            Farmers Market   
4          Sandwich Place              Beer Bar        Mexican Restaurant   
7          Baseball Field    Italian Restaurant                      Café   
..                    ...                   ...                       ...   
97             Steakhouse           Salad Place          Sushi Restaurant   
98                Dog Run      Doner Restaurant                Donut Shop   
99   Fast Food Restaurant           Men's Store  Mediterranean Restaurant   
100        Farmers Market            Comic Shop                      Park   
102                Bakery        Discount Store      Fast Food Restaurant   

    10th Most Common Venue  
1                  Dog Run  
2               Restaurant  
3       Falafel Restaurant  
4                     Café  
7       Chinese Restaurant  
..                     ...  
97      Seafood Restaurant  
98     Dumpling Restaurant  
99                     Pub  
100            Pizza Place  
102            Wings Joint  

[82 rows x 12 columns]

In [95]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
11  West Deane Park, Princess Gardens, Martin Grov...               3   

   1st Most Common Venue 2nd Most Common Venue        3rd Most Common Venue  \
11                Bakery           Yoga Studio  Eastern European Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
11               Dog Run      Doner Restaurant            Donut Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
11             Drugstore   Dumpling Restaurant     Electronics Store   

   10th Most Common Venue  
11    Filipino Restaurant

In [96]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
57                                    Humberlea, Emery               4   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...               4   

    1st Most Common Venue 2nd Most Common Venue        3rd Most Common Venue  \
57         Baseball Field           Yoga Studio  Eastern European Restaurant   
101        Baseball Field           Yoga Studio  Eastern European Restaurant   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
57                Dog Run      Doner Restaurant            Donut Shop   
101               Dog Run      Doner Restaurant            Donut Shop   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
57              Drugstore   Dumpling Restaurant     Electronics Store   
101             Drugstore   Dumpling Restaurant     Electronics Store   

    10th Most Common Venue  
57     Filipino Restaurant  
101    Filipino Restaurant

In [97]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
50  Humber Summit               5           Pizza Place          Intersection   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
50                   Gym           Yoga Studio   Dumpling Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
50   Distribution Center               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
50            Donut Shop              Drugstore

In [98]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 6, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
6                                     Malvern, Rouge               6   
56  Del Ray, Mount Dennis, Keelsdale and Silverthorn               6   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
6   Fast Food Restaurant                 Diner        Farmers Market   
56  Fast Food Restaurant        Discount Store        Sandwich Place   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
6     Falafel Restaurant           Event Space  Ethiopian Restaurant   
56    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
6            Escape Room     Electronics Store  Eastern European Restaurant   
56           Escape Room     Electronics Store         Colombian Restaurant   

   10th Most Common Venue  
6     Dumpling Restaurant  
56    Dumpling Restaurant

In [99]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 7, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  \
32  Scarborough Village               7            Playground   

   2nd Most Common Venue    3rd Most Common Venue 4th Most Common Venue  \
32                   Spa  Comfort Food Restaurant            Comic Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32        Farmers Market    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32  Ethiopian Restaurant           Escape Room      Electronics Store

In [100]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 8, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
12  Rouge Hill, Port Union, Highland Creek               8   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
12         Moving Target                   Bar   Dumpling Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
12   Distribution Center               Dog Run      Doner Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
12            Donut Shop             Drugstore           Yoga Studio   

   10th Most Common Venue  
12         Discount Store

In [101]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 9, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
51  Cliffside, Cliffcrest, Scarborough Village West               9   

   1st Most Common Venue 2nd Most Common Venue        3rd Most Common Venue  \
51                 Motel   American Restaurant  Eastern European Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
51   Distribution Center               Dog Run      Doner Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
51            Donut Shop             Drugstore   Dumpling Restaurant   

   10th Most Common Venue  
51            Yoga Studio

## Examine Clusters - Verdict

The clustering could be better. Of the 10 clusters, there are 6 that contain only a single element, and 2 that contain only two neighbourhoods. To be able to employ an entropy metric to numerically evaluate the relative entropy of different numbers of clusters, we would need to have true labels against which to evaluate - which we don't.

That aside, the distinguishing factors behind each of the labels appears to be:
0: "Park"
1: "Coffee Shop"
2: Difficult to say, but probably "Coffee Shop & Yoga Studio"
3: "Bakery"
4: These are basically identical, but "Baseball Field & Yoga Studio"
5: "Pizza Place"
6: "Fast Food Restaurant"
7: "Playground"
8: "Moving Target"
9: "Motel"